plugin.audio.mixcloud/lib/listbuilder.py

```diff
@@ -0,0 +1,289 @@
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+KODI Mixcloud Plugin is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with KODI Mixcloud Plugin.  If not, see <http://www.gnu.org/licenses/>.
+'''
+
+
+
+import sys
+import xbmc
+import xbmcgui
+import xbmcplugin
+from datetime import datetime
+from .mixcloud import MixcloudInterface
+from .utils import Utils
+from .history import History
+from .resolver import ResolverBuilder
+from .base import BaseBuilder, BaseListBuilder, QueryListBuilder, BaseList
+from .lang import Lang
+
+
+
+# main menu
+class MainBuilder(BaseListBuilder):
+
+    def buildItems(self):
+        Utils.log('MainBuilder.buildItems()')
+        if MixcloudInterface().profileLoggedIn():
+            self.addFolderItem({'title' : Lang.FOLLOWINGS}, {'mode' : 'playlists', 'key' : '/me/following/'}, Utils.getIcon('nav/kodi_highlight.png'))
+            self.addFolderItem({'title' : Lang.FOLLOWERS}, {'mode' : 'playlists', 'key' : '/me/followers/'}, Utils.getIcon('nav/kodi_highlight.png'))
+            self.addFolderItem({'title' : Lang.FAVORITES}, {'mode' : 'cloudcasts', 'key' : '/me/favorites/'}, Utils.getIcon('nav/kodi_favorites.png'))
+            self.addFolderItem({'title' : Lang.UPLOADS}, {'mode' : 'cloudcasts', 'key' : '/me/cloudcasts/'}, Utils.getIcon('nav/kodi_uploads.png'))
+            self.addFolderItem({'title' : Lang.PLAYLISTS}, {'mode' : 'playlists', 'key' : '/me/playlists/'}, Utils.getIcon('nav/kodi_playlists.png'))
+            self.addFolderItem({'title' : Lang.LISTEN_LATER}, {'mode' : 'cloudcasts', 'key' : '/me/listen-later/'}, Utils.getIcon('nav/kodi_listenlater.png'))
+        else:
+            self.addFolderItem({'title' : Lang.PROFILE}, {'mode' : 'profile', 'key' : 'login'}, Utils.getIcon('nav/kodi_profile.png'))
+        self.addFolderItem({'title' : Lang.CATEGORIES}, {'mode' : 'playlists', 'key' : '/categories/'}, Utils.getIcon('nav/kodi_categories.png'))
+        self.addFolderItem({'title' : Lang.HISTORY}, {'mode' : 'playhistory', 'offset' : 0, 'offsetex' : 0}, Utils.getIcon('nav/kodi_history.png'))
+        self.addFolderItem({'title' : Lang.SEARCH}, {'mode' : 'search'}, Utils.getIcon('nav/kodi_search.png'))
+        return 0
+
+
+
+# cloudcasts menu
+class CloudcastsBuilder(BaseListBuilder):
+
+    def buildItems(self):
+        Utils.log('CloudcastsBuilder.buildItems()')
+        xbmcplugin.setContent(self.plugin_handle, 'songs')
+        cloudcasts = MixcloudInterface().getList(self.key, {'offset' : self.offset})
+        for cloudcast in cloudcasts.items:
+            # user aborted
+            if xbmc.Monitor().abortRequested():

```

Comment:
It would be better to have a single `xbmc.Monitor()` instance here, and call `abortRequested()` on that single instance, instead of creating a lot of new `xbmc.Monitor()` instances.